File: mobile/tools/chunk_parity.ts
// Chunk-parity check (M1 verification). Reads extract.json + py_chunks.json
// produced by chunk_parity.py, runs the ported TS chunker on the same input, and
// asserts the output matches the Python chunker exactly (count, page_number,
// token_count, text). Run with: npx tsx chunk_parity.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { chunkPages } from "../app/src/chunk";
import { PageExtract } from "../app/src/types";

const here = import.meta.dirname;

type PyChunk = { page_number: number; text: string; token_count: number };

function checkCase(name: string): string[] {
  const pages: PageExtract[] = JSON.parse(readFileSync(join(here, `${name}_extract.json`), "utf8"));
  const py: PyChunk[] = JSON.parse(readFileSync(join(here, `${name}_chunks.json`), "utf8"));
  const ts = chunkPages(pages);

  const problems: string[] = [];
  if (ts.length !== py.length)
    problems.push(`[${name}] chunk count: TS ${ts.length} vs Python ${py.length}`);
  const n = Math.min(ts.length, py.length);
  for (let i = 0; i < n; i++) {
    if (ts[i].page_number !== py[i].page_number)
      problems.push(`[${name}] chunk ${i} page_number: TS ${ts[i].page_number} vs PY ${py[i].page_number}`);
    if (ts[i].token_count !== py[i].token_count)
      problems.push(`[${name}] chunk ${i} token_count: TS ${ts[i].token_count} vs PY ${py[i].token_count}`);
    if (ts[i].text !== py[i].text)
      problems.push(
        `[${name}] chunk ${i} text differs:\n  TS: ${JSON.stringify(ts[i].text)}\n  PY: ${JSON.stringify(py[i].text)}`,
      );
  }
  if (!problems.length) console.log(`  [${name}] ${ts.length} chunks identical (page, tokens, text).`);
  return problems;
}

const problems = [...checkCase("sample"), ...checkCase("synthetic")];
if (problems.length) {
  console.error("CHUNK PARITY FAILED:\n" + problems.join("\n"));
  process.exit(1);
}
console.log("CHUNK PARITY PASSED.");

File: mobile/tools/chunk_parity.py
"""Chunk-parity fixture generator (M1 verification).

Runs the WEB chunker (backend/app/ingestion/chunk.py) on a sample PDF and writes:
  - extract.json   the per-page blocks (input for the TS side)
  - py_chunks.json the Python chunk output (page_number, text, token_count)

The TS side (chunk_parity.ts) reads extract.json, runs the ported chunkPages,
and asserts its output matches py_chunks.json. Run this first, then the TS check.

Usage:
  ../../backend/.venv/bin/python chunk_parity.py [sample.pdf]
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

# Make the backend package importable.
ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "backend"))

from app.ingestion.chunk import chunk_pages  # noqa: E402
from app.ingestion.extract import PageExtract, Block, extract_pdf  # noqa: E402

OUT = Path(__file__).parent
DEFAULT_PDF = ROOT / "mobile" / "extract-service" / "samples" / "sample.pdf"


def _synthetic_pages() -> list[PageExtract]:
    """Pages that exercise the branches the tiny sample PDF doesn't: multi-chunk
    packing + overlap between text chunks, and an oversized single paragraph that
    must be window-split. Word tokens are intentionally distinct so any drift in
    boundaries/overlap shows up as a text mismatch."""
    many_paras = "\n\n".join(
        " ".join(f"w{p}-{i}" for i in range(80)) for p in range(20)
    )  # 20 paras x 80 words -> spans several 512-token chunks with overlap
    huge_para = " ".join(f"x{i}" for i in range(1300))  # one >512-token paragraph
    return [
        PageExtract(page_number=1, blocks=[Block(kind="text", content=many_paras)]),
        PageExtract(
            page_number=2,
            blocks=[
                Block(kind="text", content=huge_para),
                Block(kind="table", content="| A | B |\n| --- | --- |\n| 1 | 2 |"),
            ],
        ),
    ]


def _dump(name: str, pages: list[PageExtract]) -> int:
    extract = [
        {
            "page_number": p.page_number,
            "blocks": [{"kind": b.kind, "content": b.content} for b in p.blocks],
        }
        for p in pages
    ]
    chunks = [
        {"page_number": c.page_number, "text": c.text, "token_count": c.token_count}
        for c in chunk_pages(pages)
    ]
    (OUT / f"{name}_extract.json").write_text(json.dumps(extract, ensure_ascii=False, indent=2))
    (OUT / f"{name}_chunks.json").write_text(json.dumps(chunks, ensure_ascii=False, indent=2))
    print(f"  {name}: {len(extract)} pages -> {len(chunks)} chunks")
    return len(chunks)


def main() -> int:
    pdf = Path(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT_PDF
    print("wrote parity fixtures:")
    _dump("sample", extract_pdf(pdf))
    _dump("synthetic", _synthetic_pages())
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
